cirq/sim/wave_function.py

```diff
@@ -345,36 +357,137 @@ def to_valid_state_vector(
     Raises:
         ValueError if the state is not valid or num_qubits != len(qid_shape).
     """
+
+    # Check shape.
+    if num_qubits is None and qid_shape is None:
+        raise ValueError('Must specify `num_qubits` or `qid_shape`.')
     if qid_shape is None:
         qid_shape = (2,) * num_qubits
+    else:
+        qid_shape = tuple(qid_shape)
+    if num_qubits is None:
+        num_qubits = len(qid_shape)
     if num_qubits != len(qid_shape):
         raise ValueError('num_qubits != len(qid_shape). num_qubits is <{!r}>. '
                          'qid_shape is <{!r}>.'.format(num_qubits, qid_shape))
-    if isinstance(state_rep, np.ndarray):
-        if len(state_rep) != np.prod(qid_shape, dtype=int):
-            raise ValueError(
-                'initial state was of size {} '
-                'but expected state for {} qubits with qid shape {}'.format(
-                    len(state_rep), num_qubits, qid_shape))
-        state = state_rep.copy()
-    elif isinstance(state_rep, int):
-        if state_rep < 0:
-            raise ValueError('initial_state must be positive')
-        elif state_rep >= np.prod(qid_shape, dtype=int):
-            raise ValueError(
-                'initial state was {} but expected state for {} qubits'.format(
-                    state_rep, num_qubits))
-        else:
-            state = linalg.one_hot(shape=np.prod(qid_shape, dtype=int),
-                                   dtype=dtype,
-                                   index=state_rep)
-    else:
-        raise TypeError('initial_state was not of type int or ndarray')
-    validate_normalized_state(state,
+
+    tensor = _state_like_to_state_tensor(state_like=state_rep,
+                                         qid_shape=qid_shape,
+                                         dtype=dtype,
+                                         atol=atol)
+    return tensor.reshape(tensor.size)  # Flatten.
+
+
+def _state_like_to_state_tensor(*, state_like: 'cirq.STATE_VECTOR_LIKE',
+                                qid_shape: Tuple[int, ...],
+                                dtype: Type[np.number], atol) -> np.ndarray:
+
+    if isinstance(state_like, int):
```

Comment:
See #2375.

Reply:
In this case I really do want it to be an integer and not a generic number such as a float. I suppose numpy ints would also be acceptable, though.